File: app/src/main/java/com/w3engineers/unicef/telemesh/ui/createuser/CreateUserActivity.java
package com.w3engineers.unicef.telemesh.ui.createuser;

import android.Manifest;
import android.annotation.SuppressLint;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import android.app.ProgressDialog;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import android.text.TextUtils;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import com.w3engineers.unicef.telemesh.R;
import com.w3engineers.unicef.telemesh.data.helper.constants.Constants;
import com.w3engineers.unicef.telemesh.data.provider.ServiceLocator;
import com.w3engineers.unicef.telemesh.databinding.ActivityCreateUserBinding;
import com.w3engineers.unicef.telemesh.ui.main.MainActivity;
import com.w3engineers.unicef.telemesh.ui.selectaccount.SelectAccountActivity;
import com.w3engineers.unicef.util.base.ui.BaseActivity;
import com.w3engineers.unicef.util.helper.CommonUtil;
import com.w3engineers.unicef.util.helper.uiutil.UIHelper;

import java.util.List;

/*
 * ============================================================================
 * Copyright (C) 2019 W3 Engineers Ltd - All Rights Reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 * ============================================================================
 */

public class CreateUserActivity extends BaseActivity implements View.OnClickListener {

    private ActivityCreateUserBinding mBinding;
    public static int INITIAL_IMAGE_INDEX = -1;
    private CreateUserViewModel mViewModel;

    public static CreateUserActivity sInstance;
    private boolean isNeedToImportWallet;
    private boolean isWalletExists;
    private ProgressDialog progressDialog;

    @NonNull
    public static String IMAGE_POSITION = "image_position";


    @Override
    protected int getLayoutId() {
        return R.layout.activity_create_user;
    }

    @Override
    protected int statusBarColor() {
        return R.color.colorPrimaryDark;
    }

    @SuppressLint("CheckResult")
    @Override
    protected void startUI() {

        mBinding = (ActivityCreateUserBinding) getViewDataBinding();

        isNeedToImportWallet = getIntent().getBooleanExtra(Constants.IntentKeys.IMPORT_WALLET, false);

        isWalletExists = getIntent().getBooleanExtra(Constants.IntentKeys.WALLET_EXISTS, false);

        setTitle(getString(R.string.create_user));
        mViewModel = getViewModel();

        setClickListener(mBinding.imageViewBack, mBinding.buttonSignup);

        UIHelper.hideKeyboardFrom(this, mBinding.editTextFirstName);
        UIHelper.hideKeyboardFrom(this, mBinding.editTextLastName);

        mViewModel.firstNameChangeLiveData.observe(this, this::nextButtonControl);
        mViewModel.lastNameChangeLiveData.observe(this, this::nextButtonControl);


        mViewModel.firstNameEditControl(mBinding.editTextFirstName);
        mViewModel.lastNameEditControl(mBinding.editTextLastName);

        sInstance = this;

        mBinding.editTextFirstName.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_NEXT) {
                nextAction();
                return true;
            }
            return false;
        });
    }

    private void nextButtonControl(String nameText) {

        updateImageNameField();


        String firstName = mBinding.editTextFirstName.getText().toString();
        String lastName = mBinding.editTextLastName.getText().toString();

        if (!TextUtils.isEmpty(firstName)
                && !TextUtils.isEmpty(lastName)
                && firstName.length() >= Constants.DefaultValue.MINIMUM_TEXT_LIMIT
                && lastName.length() >= Constants.DefaultValue.MINIMUM_TEXT_LIMIT) {

            mBinding.buttonSignup.setBackgroundResource(R.drawable.ractangular_gradient);
            mBinding.buttonSignup.setTextColor(getResources().getColor(R.color.white));
            //mBinding.buttonSignup.setClickable(true);
        } else {

            mBinding.buttonSignup.setBackgroundResource(R.drawable.ractangular_white);
            mBinding.buttonSignup.setTextColor(getResources().getColor(R.color.new_user_button_color));
            //mBinding.buttonSignup.setClickable(false);
        }
    }

    private void updateImageNameField() {
        String firstName = mBinding.editTextFirstName.getText().toString();
        String lastName = mBinding.editTextLastName.getText().toString();

        String finalText = "";


        if (!TextUtils.isEmpty(firstName)) {

            finalText = String.valueOf(firstName.charAt(0));
        }
        if (!TextUtils.isEmpty(lastName)) {

            finalText += String.valueOf(lastName.charAt(0));
        }

        if (TextUtils.isEmpty(finalText)) {
            mBinding.textViewImageName.setVisibility(View.GONE);

            mBinding.imageProfileBackground.setVisibility(View.VISIBLE);
            mBinding.imageProfile.setVisibility(View.VISIBLE);
        } else {
            mBinding.textViewImageName.setVisibility(View.VISIBLE);

            mBinding.imageProfileBackground.setVisibility(View.INVISIBLE);
            mBinding.imageProfile.setVisibility(View.GONE);
        }

        mBinding.textViewImageName.setText(finalText);
    }

    private CreateUserViewModel getViewModel() {
        return ViewModelProviders.of(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                return (T) ServiceLocator.getInstance().getCreateUserViewModel(getApplication());
            }
        }).get(CreateUserViewModel.class);
    }

    protected void requestMultiplePermissions() {

        Dexter.withActivity(this)
                .withPermissions(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE/*,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        Manifest.permission.ACCESS_FINE_LOCATION*/)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {

                        if (report.areAllPermissionsGranted()) {
                            goNext();
                        }

                        // check for permanent denial of any permission
                        if (report.isAnyPermissionPermanentlyDenied()) {
                            CommonUtil.showPermissionPopUp(CreateUserActivity.this);
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(
                            List<PermissionRequest> permissions, PermissionToken token) {
                        token.continuePermissionRequest();
                    }
                }).withErrorListener(error -> requestMultiplePermissions()).onSameThread().check();
    }

    @Override
    public void onClick(@NonNull View view) {
        super.onClick(view);

        int id = view.getId();

        if (id == R.id.button_signup) {
            nextAction();
        } else if (id == R.id.image_view_back) {
            finish();
        }

    }

    private void nextAction() {

        saveData();
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        sInstance = null;
    }

    private void saveData() {

        if (CommonUtil.isValidName(mBinding.editTextFirstName.getText().toString(), this, true)
                && CommonUtil.isValidName(mBinding.editTextLastName.getText().toString(), this, false)) {
            requestMultiplePermissions();
        }
    }

    protected void goNext() {
        if (mViewModel.storeData(mBinding.editTextFirstName.getText() + "",
                mBinding.editTextLastName.getText() + "")) {
            toggleProgressDialog(true);
            if (isWalletExists) {

                // Start mesh, and goto home page

                gotoHomePage();

            } else {
                if (isNeedToImportWallet) {

                    gotoHomePage();

                } else {
                    mViewModel.launchWalletPage(isNeedToImportWallet);
                }
            }

            /*Intent intent = new Intent(CreateUserActivity.this, MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(intent);
            finish();*/
        }
    }

    private void toggleProgressDialog(boolean needToShow) {
        if (needToShow) {
            progressDialog = new ProgressDialog(this);
            progressDialog.setCancelable(true);
            progressDialog.setMessage("Please wait...");
            progressDialog.show();
        } else {
            if (progressDialog != null) {
                progressDialog.dismiss();
                progressDialog = null;
            }
        }
    }

    private void gotoHomePage() {

        Intent intent = new Intent(this, MainActivity.class);
        intent.putExtra(Constants.IntentKeys.IS_MESH_START, true);
        intent.setAction(MainActivity.class.getName());
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
        finish();

        if (SelectAccountActivity.instance != null) {
            SelectAccountActivity.instance.finish();
        }

        //RmDataHelper.getInstance().startMesh();
    }
}

File: app/src/main/java/com/w3engineers/unicef/telemesh/ui/createuser/CreateUserActivity.kt
package com.w3engineers.unicef.telemesh.ui.createuser

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.text.Html
import android.text.TextUtils
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import com.karumi.dexter.Dexter
import com.karumi.dexter.MultiplePermissionsReport
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.DexterError
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.multi.MultiplePermissionsListener
import com.w3engineers.ext.strom.application.ui.base.BaseActivity
import com.w3engineers.unicef.telemesh.R
import com.w3engineers.unicef.telemesh.data.helper.constants.Constants
import com.w3engineers.unicef.telemesh.data.provider.ServiceLocator
import com.w3engineers.unicef.telemesh.ui.chooseprofileimage.ProfileImageActivity
import com.w3engineers.unicef.telemesh.ui.importwallet.ImportWalletActivity
import com.w3engineers.unicef.telemesh.ui.main.MainActivity
import com.w3engineers.unicef.telemesh.ui.security.SecurityActivity
import com.w3engineers.unicef.util.helper.CommonUtil
import com.w3engineers.unicef.util.helper.WalletAddressHelper
import com.w3engineers.unicef.util.helper.uiutil.UIHelper
import com.w3engineers.walleter.wallet.WalletService
import kotlinx.android.synthetic.main.activity_create_user.*
/*
* ============================================================================
* Copyright (C) 2019 W3 Engineers Ltd - All Rights Reserved.
* Unauthorized copying of this file, via any medium is strictly prohibited
* Proprietary and confidential
* ============================================================================
*/
class CreateUserActivity : BaseActivity(), View.OnClickListener {

    // private var mViewModel: CreateUserViewModel? = null
    // private var mBinding: ActivityCreateUserBinding? = null
    // private var mPassword: String? = null

    // lateinit: Use this keyword when you know the value of a property will not be null once it is initialized.
    //private lateinit var mBinding: ActivityCreateUserBinding
    private lateinit var mViewModel: CreateUserViewModel
    private lateinit var mPassword: String

    private val PROFILE_IMAGE_REQUEST = 1

    // Kotlin can infer the type from the assignment, so no boolean type is required to indicate
    private var isLoadAccount = false


    // to expose the field to Java class @JvmField annotation is used
    companion object {
        @JvmField var INITIAL_IMAGE_INDEX = -1
        @JvmField var sInstance: CreateUserActivity? = null
        @JvmField var IMAGE_POSITION = "image_position"
    }


    override fun getLayoutId(): Int {
        return R.layout.activity_create_user
    }

    override fun statusBarColor(): Int {
        return R.color.colorPrimaryDark
    }

    @SuppressLint("CheckResult")
    override fun startUI() {
        //mBinding = viewDataBinding as ActivityCreateUserBinding
        setTitle(getString(R.string.create_user))
        mViewModel = createUserViewModel

        setClickListener(imageViewBack)
        parseIntent()

        UIHelper.hideKeyboardFrom(this, editTextName)
        imageViewCamera.setOnClickListener(this)
        buttonSignup.setOnClickListener(this)
        imageProfile.setOnClickListener(this)

        editTextName.maxCharacters = Constants.DefaultValue.MAXIMUM_TEXT_LIMIT
        editTextName.minCharacters = Constants.DefaultValue.MINIMUM_TEXT_LIMIT

        mViewModel.textChangeLiveData.observe(this, { nameText: String -> nextButtonControl(nameText) })

        mViewModel.textEditControl(editTextName)
        sInstance = this

        editTextName.setOnEditorActionListener { v, actionId, event ->
            if (actionId === EditorInfo.IME_ACTION_NEXT) {
                nextAction()
                return@setOnEditorActionListener true
            }
            false
        }
    }

    private fun nextButtonControl(nameText: String) {

        if (!TextUtils.isEmpty(nameText) &&
                nameText.length >= Constants.DefaultValue.MINIMUM_TEXT_LIMIT) {
            buttonSignup.setBackgroundResource(R.drawable.ractangular_gradient)
            buttonSignup.setTextColor(resources.getColor(R.color.white))
            //mBinding.buttonSignup.setClickable(true);
        } else {
            buttonSignup.setBackgroundResource(R.drawable.ractangular_white)
            buttonSignup.setTextColor(resources.getColor(R.color.new_user_button_color))
            //mBinding.buttonSignup.setClickable(false);
        }
    }

    private val createUserViewModel: CreateUserViewModel
         get() = ViewModelProviders.of(this, object : ViewModelProvider.Factory {
            override fun <T : ViewModel?> create(modelClass: Class<T>): T {
                return ServiceLocator.getInstance().getCreateUserViewModel(application) as T
            }
        }).get(CreateUserViewModel::class.java)

    protected fun requestMultiplePermissions() {
        Dexter.withActivity(this)
                .withPermissions(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        Manifest.permission.ACCESS_FINE_LOCATION)
                .withListener(object : MultiplePermissionsListener {
                    override fun onPermissionsChecked(report: MultiplePermissionsReport) {
                        if (report.areAllPermissionsGranted()) {
                            goNext()
                        }

                        // check for permanent denial of any permission
                        if (report.isAnyPermissionPermanentlyDenied) {
                            CommonUtil.showPermissionPopUp(this@CreateUserActivity)
                        }
                    }

                    override fun onPermissionRationaleShouldBeShown(
                            permissions: List<PermissionRequest>, token: PermissionToken) {
                        token.continuePermissionRequest()
                    }
                }).withErrorListener { error: DexterError? -> requestMultiplePermissions() }.onSameThread().check()
    }

    override fun onClick(view: View) {
        super.onClick(view)
        val id = view.id
        when (id) {
            R.id.buttonSignup -> nextAction()
            R.id.imageProfile, R.id.imageViewCamera -> {
                val intent = Intent(this, ProfileImageActivity::class.java)
                intent.putExtra(IMAGE_POSITION, mViewModel.imageIndex)
                startActivityForResult(intent, PROFILE_IMAGE_REQUEST)
            }
            R.id.imageViewBack -> finish()
        }
    }

    private fun nextAction() {
        if (isLoadAccount) {
            saveData()
        } else {
            goToPasswordPage()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data != null && requestCode == PROFILE_IMAGE_REQUEST && resultCode == RESULT_OK) {
            mViewModel.imageIndex = data.getIntExtra(IMAGE_POSITION, INITIAL_IMAGE_INDEX)
            val id = resources.getIdentifier(Constants.drawables.AVATAR_IMAGE + mViewModel.imageIndex, Constants.drawables.AVATAR_DRAWABLE_DIRECTORY, packageName)
            imageProfile.setImageResource(id)


            nextButtonControl(editTextName.text.toString())

        }
    }

    override fun onDestroy() {
        super.onDestroy()
        sInstance = null
    }

    private fun saveData() {

        // Create an immutable variable that stores a lambda expression called notEmpty.
        // The lambda expression takes a TextView as a parameter (EditText inherits from TextView)
        // and returns a Boolean:
        // If a lambda expression only has a single parameter,
        // it can be omitted and replaced with the it keyword.

        // The pattern is input -> output, however, if the code returns no value we use the type Unit:
        val notEmpty: (TextView) -> Boolean = {textView ->  textView.text.isEmpty()}
        if(notEmpty(editTextName)){
            if (CommonUtil.isValidName(editTextName.text.toString(), this)) {
                requestMultiplePermissions()
            }
        }

    }

    private fun parseIntent() {
        val intent = intent
        if (intent.hasExtra(Constants.IntentKeys.PASSWORD)) {
            isLoadAccount = true
            mPassword = intent.getStringExtra(Constants.IntentKeys.PASSWORD)
        } else {
            if (WalletService.getInstance(this).isWalletExists) {
                showWarningDialog()
            }
        }
    }

    protected fun goNext() {
        if (mViewModel.storeData(editTextName.text.toString() + "", mPassword)) {
            val intent = Intent(this@CreateUserActivity, MainActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
            startActivity(intent)
            finish()
        }
    }

    private fun goToPasswordPage() {
        if (CommonUtil.isValidName(editTextName.text.toString(), this)) {
            val intent = Intent(this@CreateUserActivity, SecurityActivity::class.java)
            intent.putExtra(Constants.IntentKeys.USER_NAME, editTextName.text.toString() + "")
            intent.putExtra(Constants.IntentKeys.AVATAR_INDEX, mViewModel.imageIndex)
            startActivity(intent)
        }
    }

    private fun showWarningDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setCancelable(false)
        builder.setTitle(Html.fromHtml("<b>" + getString(R.string.alert_title_text) + "</b>"))
        builder.setMessage(WalletAddressHelper.getWalletSpannableString(this).toString())
        builder.setPositiveButton(Html.fromHtml("<b>" + getString(R.string.button_postivive) + "<b>")) { dialog, arg1 ->
            startActivity(Intent(this@CreateUserActivity, ImportWalletActivity::class.java))
            finish()
        }
        builder.setNegativeButton(Html.fromHtml("<b>" + getString(R.string.negative_button) + "<b>")) { dialog, arg1 -> }
        builder.create()
        builder.show()
    }


}